internal/keystore/entrust/keycontrol.go

```diff
@@ -0,0 +1,575 @@
+		Path       = "/vault/1.0/ListSecretIds/"
+		VaultToken = "X-Vault-Auth"
+	)
+	type Token struct {
+		BoxID      string `json:"box_id"`
+		ContinueAt string `json:"next_ctx"`
+	}
+	type Request struct {
+		BoxID     string `json:"box_id"`
+		N         int    `json:"max_items"`
+		NextToken string `json:"next_token,omitempty"` // Must be omitted if empty
+	}
+	type Response struct {
+		Secrets []struct {
+			Name    string `json:"name"`
+			Expired bool   `json:"expired"`
+		}
+		NextToken string `json:"next_token"`
+	}
+
+	var token string
+	if prefix != "" {
+		b, err := json.Marshal(Token{
+			BoxID:      kc.config.BoxID,
+			ContinueAt: prefix,
+		})
+		if err != nil {
+			return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+		}
+		token = base64.StdEncoding.EncodeToString(b)
+	}
+	body, err := json.Marshal(Request{
+		BoxID:     kc.config.BoxID,
+		N:         n,
+		NextToken: token,
+	})
+	if err != nil {
+		return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+	}
+
+	url, err := url.JoinPath(kc.config.Endpoint, Path)
+	if err != nil {
+		return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+	}
+	req, err := http.NewRequestWithContext(ctx, Method, url, xhttp.RetryReader(bytes.NewReader(body)))
+	if err != nil {
+		return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+	}
+	req.ContentLength = int64(len(body))
+	req.Header.Set(VaultToken, *kc.token.Load())
+
+	resp, err := kc.client.Do(req)
+	if err != nil {
+		return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, "", parseErrorResponse(resp)
+	}
+
+	var response Response
+	if err := json.NewDecoder(mem.LimitReader(resp.Body, 10*mem.MB)).Decode(&response); err != nil {
+		return nil, "", fmt.Errorf("keycontrol: failed to list keys: %v", err)
+	}
+	names := make([]string, len(response.Secrets))
+	for _, secret := range response.Secrets {
+		if !secret.Expired {
+			names = append(names, secret.Name)
+		}
+	}
+	if response.NextToken != "" {
+		rawToken, err := base64.StdEncoding.DecodeString(response.NextToken)
+		if err != nil {
+			return nil, "", fmt.Errorf("keycontrol: failed to list keys: invalid continue token: %v", err)
+		}
+
+		var token Token
+		if err = json.Unmarshal(rawToken, &token); err != nil {
+			return nil, "", fmt.Errorf("keycontrol: failed to list keys: invalid continue token: %v", err)
+		}
+		prefix = token.ContinueAt
+	} else {
+		prefix = ""
+	}
+	return names, prefix, nil
+}
+
+func (kc *KeyControl) refreshToken(ctx context.Context, renew time.Duration) {
+	if renew/2 <= 0 {
+		renew = 5 * time.Minute
+	}
+	timer := time.NewTimer(renew / 2)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			token, expiresAt, err := renewToken(ctx, kc.client.Transport, kc.config.Endpoint, *kc.token.Load())
+			if err != nil {
+				log.Default().Printf("keycontrol: failed to renew auth token: %v", err)
```

Comment:
Should we mention in log that, due to failure in renew of token, its doing re-login?

Reply:
I think not needed. We try to login again and in case of failure also log that we failed to authenticate using username/password.
Also I try to avoid adding non-error log since I want to rework the logging (eventually move to e.g. log/slog) and have separate logging for errors (HTTP, application etc), audit, traces, a.s.o - to avoid mixing everything together...